### ISSUE TITLE
Basic Queries

### ISSUE THREAD

**Issue by AyushmanTripathy** [AyushmanTripathy/language#1]
basic questions about the projects

1. language to be used?
2. lisence?

can we start with this "language" name and change it afterwards ?

**Comment by dumblob** [comment]
> 1. language to be used?

It's up to you. I'd actually split this question into two:

a) Language to be used for the default compiler.
b) Language to be used as the target to compile to (assembly? Go? C? [V](https://github.com/vlang/v )? [Nim](https://nim-lang.org/ )? [Basil](https://github.com/basilTeam/basil )? Python? ...).

For (b) I'd go with something which compiles extremely quickly (i.e. either V or Go or Basil or C or anything interpreted like Python) and which is multiplatform in a sense "it's dead easy to install even by an unexperienced user". Note, I'd discourage C++ because it's utterly slow and can't be made fast. I'd also discourage targetting some low-level language like assembly because it'd derail the focus from actually designing a new language to dealing with many shortcomings (register allocation, stack memory management, heap memory management, alignment, padding, no emulation of 64bit integers on 32bit platforms, and tons of other irrelevant stuff...).

Other concerns as "speed" etc. are IMHO irrelevant as generally stream-oriented programming is very fast per se (the difference is only in latency - which can admittedly get very high in slow languages).

But mind that changing the target language is much lower burden than changing the language for (a). So we don't have to be that careful IMHO. I myself would probably start with V as target language, but it's up to you. As language for (a) I don't have any particular favourite. Maybe it could be worth it to use the same language for both (a) and (b) to ease the mental burden, but again, it's up to you.

> 2. lisence?

It can be relicensed any time. So I'd first go with MIT (or BSD) and if any issues shall arise, you can change it any time.

> can we start with this "language" name and change it afterwards ?

Sure, the name can change any time even several times. Though I'd recommend starting with some more specific name to ease the search for this repo. Maybe "Razite" (borrowing from "raz" which in turn borrows from Old Norse rás‎ ("current, channel")):

>### Razite
>A simple safe pipe programming language with full reactivity and manycore support.

(see other cool names on https://www.samcodes.co.uk/project/markov-namegen/ )

**Comment by AyushmanTripathy** [comment]
lets use vlang for both (a) and (b). i will learn vlang on the way.
now where should we start?

**Comment by dumblob** [comment]
I'd start with considering "repurposing" your effort for study purposes. I'd strangly recommend you finding some subject(s) of your studies where you could present Razite e.g. as a seminal project or as your bachelor thesis topic or such.

You'd have more stable motivation to work on it, you'd have much clearer boundaries "what's necessary to achieve soon and what's nice to have later", you'd also get more professionals looking over your shoulders which is always a good thing.

----------

Now, to the technical part of your question. First I'd prepare a set of example programs in the Razite syntax and you'd take the simpliest example and implement a compiler which compiles it. Then you'd take the second example (which will slightly raise the difficulty bar) and change the compiler to handle the new example as well while maintaining the first example working. Then you'd take a third example etc.

This is called [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development#Test-driven_development_cycle ) and is very efficient for kicking off such projects where you know decently well up front how the result shall look like. The tests should be fully automated (later using GitHub automation). For development I highly recommend using `v watch` and `v test` both from a `Makefile`.

Let's create first few examples (in this order):

#### 1. test

```razite
"Hello world!"
```
The compiler shall support two modes - one "production level compilation" which will be strict in a sense that it **won't compile** in case of any warning or error. The other mode will be a debug mode and will be the default (it'll also be used for interactive REPL if we'll decide we want a REPL at some point in the future - now it's not a priority).

The production mode could be turned on by `razite prod file app.raz` (the `filein` argument expects the following argument to be an existing UTF-8 text file with Razite syntax) and will produce `app.exe` in the same directory. The debug mode will work with `razite run app.raz` which will produce `/some/temporary/location/af86fbc394ec.exe` (randomly named executable in a temporary location) and will immediately run the compiled binary.

In debug mode this example will create an unnamed "producer node" which will produce exactly one value - a string `Hello world!`. Because there is no "consumer node" an unnamed "consumer node outputting to stderr" will be created (partially simulating `| autoview` from [nushell](https://www.nushell.sh/book/command_reference.html )) and the output of the unnamed producer node piped to the consumer node which will print it to stderr. This app will run forever (has to be killed by sigterm/Ctrl+c) and "do nothing" (the consumer will wait for more data but there will be no more data).

In production mode this won't compile at all because we would need to make `autoview` explicit (i.e. `"Hello world!" autoview`). But that's fine - most of our tests from the upcoming weeks will not compile in production mode yet :wink:.

#### 2. test

```razite
rand periodically
```

This is a more complicated example because `rand` command outputs infinite number of random numbers ([decimal floats](https://www.bytereef.org/mpdecimal/ ) from the range 0.0-1.0 with uniform distribution; but let's start with binary floats to make things easier and switch to decimal floats later) in the highest possible computable rate. `periodically` is a filter which will wait for `1` second, then read one value from its input channel (to which `rand` outputted some random numbers) and pass this value in its original form further. Then again wait for `1` second etc.

And because there is no explicit consumer, an unnamed consumer node `autoview` will be created which will consume all the outputted vaules and write them to stderr. But there is a change - because only the value changed, it'll not write it to next line of stderr but overwrite the previously outputted value!

#### 3. test

```razite
rand (norm) periodically
```

Almost the same as above but with the major difference that we'll need to extend the `rand` command to become also a consumer (so far it was only a producer), then make everything in `()` into separate producers, then plumb these through a channel to `rand`. Then make `rand` accept aliases, make it check in compile-time whether the first producer from `()` is an alias called `norm` and if so, output random values with a uniform distribution. Otherwise if it's `uni` instead of `norm`, just output values with uniform distribution.

And I think this is enough for now. It'll be a long way before we'll settle everything to get to the fizzbuzz and fibonacci etc. examples :wink:.

Next examples could start dealing with "lowering" as described by "disguising" in [(1.)](https://github.com/AyushmanTripathy/pipe-script/issues/6#issuecomment-979850625 )...

**Comment by dumblob** [comment]
Btw. I might sometimes be very slow to respond, so don't worry and work on it yourself - I want it to be your language after all :wink:.

**Comment by AyushmanTripathy** [comment]
can we change (a) to javascript, because it is really difficult to find anything about vlang online ( every search shows something about golang )

**Comment by AyushmanTripathy** [comment]
syntax something like this?

```js
// this is a comment
let name = "this is a mutable var"
const name = "this is a unmutable var"

// functions
name(int a) => int {
  return a * 2
}

// conditional flow
if 10 < 1 {  
} else {
}
```

**Comment by dumblob** [comment]
> can we change (a) to javascript

Sure. I have no problem with that.

>because it is really difficult to find anything about vlang online ( every search shows something about golang )

It's a young language - but one anyway mostly needs [language doc](https://github.com/vlang/v/blob/master/doc/docs.md ) and [stdlib doc](https://modules.vlang.io/ ).

> syntax something like this?
> 
> ```js
> // this is a comment
> let name = "this is a mutable var"
> const name = "this is a unmutable var"
> 
> // functions
> name(int a) => int {
>   return a * 2
> }
> 
> // conditional flow
> if 10 < 1 {  
> } else {
> }
> ```

Sorry, but I don't understand what you mean by this code example. I thought you're going to implement a compiler of a pipe language with the syntax I outlined above in the 3 test examples in https://github.com/AyushmanTripathy/razite/issues/1#issuecomment-982065629 .

**Comment by AyushmanTripathy** [comment]
can you plz give me a code example of what you have in mind.  ( are you taking about syntax like in pipescript )

**Comment by dumblob** [comment]
Sure. The syntax I envision is what I proposed in https://github.com/AyushmanTripathy/pipe-script/issues/6#issuecomment-979850625 :

```razite
$fizzbuzz { gen (-999 999 1) filter (x >= 1 && x <= 100)
  @ branch (x % 3 == 0 && x % 5 == 0) write (stdout FizzBuzz)  # FizzBuzz alias got stringified in `write` due to "disguising" as described below
  @ else
    @ branch (x % 5) write (stdout Buzz)
    @ else
      @ branch (x % 3) write (stdout Fizz)
      @ else trash
}  # defines & immediately executes the block with no arguments
fizzbuzz  # executes the same block again

{ args filter ... ;; log (howk) } (just three arguments)  # define a block which accepts
                                                          # some streams as arguments, then execute
                                                          # this block with arguments `just` `three` `arguments`
                                                          # (note ;; is syntax sugar to delimit independent
                                                          # subgraphs executed in parallel)
{ args filter ... } ()  # this will compile fine
{ args filter ... }  # same as previous line (but with syntax sugar allowing to omit the empty list)
```

There are no functions, no variables, no mutable/immutable distinction (everything is by definition immutable), nothing like in "stop the world languages" because every single command runs in parallel (or at least concurrently) to all other commands.

I envision that everything is automatically a stream/channel (even a string `"Hello world!"` becomes automatically a stream streaming just one sample - the string *Hello world!*). Thus there is no need to think about anything - there are no restrictions - everything can be connected (piped) to anything.

Other valuable examples:
```razite
"Hello world!"  # prints Hello world! in debug mode and runs forever without doing anything else
```
```razite
rand periodically  # prints a random number in debug mode every second
```
```razite
rand (norm) periodically  # prints a random number in debug mode from a normal distribution every second
```

Does this make more sense?

**Comment by AyushmanTripathy** [comment]
i think i get it now.

**Comment by AyushmanTripathy** [comment]
can you give me more simple instruction on where to start

**Comment by dumblob** [comment]
You mean simpler than the **1. test** in https://github.com/AyushmanTripathy/razite/issues/1#issuecomment-982065629 ?

I'm afraid there is nothing much simpler than that.

Do you know test-based programming? Aka *"first make all tests fail and then slowly take the individual tests one by one and make them work"* ([TDD](https://en.wikipedia.org/wiki/Test-driven_development ))?

If you're rather a business-oriented person, you could imagine **1. test** is an [acceptance criterion](https://en.wikipedia.org/wiki/Acceptance_testing ) you received from some customer who ordered a compiler app from you :wink:. The customer doesn't care how you will develop the app, but she just cares about the acceptance criterion. You'll give her your app and she'll run the test **1. test** and if it'll do what's defined in the acceptance criterion, she'll pay you for the app you gave her. Then you both can move to the acceptance criterion **2. test** and so on...

**Comment by dumblob** [comment]
Btw. technically your compiler shall spit out a source code which will do something like this:
- create a channel (imagine a go channel) `ch1`
- create two threads (or fibers) `t1` and `t2`
- make `t1` push the string `Hello world!` into `ch1` and then make `t1` sleep forever
- make `t2` wait for any new values from `t1` and once received, just print it to stderr

**Comment by AyushmanTripathy** [comment]
sorry that you have to waste your time on me. but these (channels and threads) are way out of my league for now. i don't think i can do something on that level yet.

**Comment by dumblob** [comment]
Ah, ok. No problem for me. Should you change your mind at some point, feel free to revive this repository and let me know :wink:.

Btw. despite "threads", "channels" etc. might sound difficult to master, languages like V and Go make them piece of cake (that's another reason why I proposed using V). So feel free to check them out, it's really easy at the beginning and over the course of the time you'll learn more about them and also about the internals (you can also ask me any time if there is anything unclear :wink:).

**Comment by AyushmanTripathy** [comment]
we will surely complete project, once i learn about these topics and learn V! its been really good to talk to someone like you!
